plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -596,27 +602,118 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String key = keyFromPath(path);
+        if (recursive) {
+            boolean deleted = deletePrefix(path);
+            if (deleted) {
+                deleteObject(key + DIRECTORY_SUFFIX);
+                return true;
+            }
+            else {
+                return deleteObject(key);
             }
         }
-        catch (FileNotFoundException e) {
+        else {
+            if (path.getName().isEmpty()) {
+                return false; // bucket root does not exist
+            }
+            Iterator<ListObjectsV2Result> listingsIterator = listObjects(path, OptionalInt.of(2), true);
+            Iterator<String> objectKeysIterator = Iterators.concat(Iterators.transform(listingsIterator, TrinoS3FileSystem::keysFromListing));
+            if (objectKeysIterator.hasNext()) {
+                String directoryKey = objectKeysIterator.next();
+                if (!Objects.equals(directoryKey, key + PATH_SEPARATOR) || objectKeysIterator.hasNext()) {
+                    throw new IOException("Directory " + path + " is not empty");
+                }
+                deleteObject(directoryKey);
+            }
+            else {
+                deleteObject(key);
+            }
+            deleteObject(key + DIRECTORY_SUFFIX);
```

Comment:
This may seem out of place because it makes sense only for directories.
However, listing with the prefix "directory/" under MinIO (oposed to AWS S3) doesn't list the "directory/" object key even if it exists. I've decided to perform two `deleteObject` operations instead of trying to figure out whether I have or not a directory by retrieving the object metadata call.

Reply:
Wouldn't the MinIO issue also affect the recursive case above?

Reply:
Interestingly enough, if the directory object ( containing `/`) is created and has objects within it, it becomes in MinIO an virtual object and is not listed anymore. When deleting the objects it contains, the "base" directory dissapears as well.

Reply:
It looks like the dockerized MinIO != S3 (from behavioral perspective)

"s3-compatible storage" is a promise, a hope.